states/visited.go

```diff
@@ -2,160 +2,13 @@ package states
-	}
-	targetStr := target.StringCanonical()
-	vc.visited[targetStr] = append(vc.visited[targetStr], sts)
-	vc.visitedList = append(vc.visitedList, sts)
-	return sts, false, nil
-}
-
-// waitAllDoneAndLock acquires mu at a point when all sts are done for a particular
-// target, allowing for comparisons across the board while the lock is held.
-func (vc *VisitedCollection) waitAllDoneAndLock(ctx context.Context, target domain.Target, parentDepSub chan string) (map[string]bool, error) {
-	// Build up dependents from parentDepSub. The list needs to be complete when returning
-	// from this function for proper infinite loop detection.
-	dependents := make(map[string]bool)
-	// wait all done & lock loop
-	prevLenList := 0
-	for {
-		vc.mu.Lock()
-		list := append([]*SingleTarget{}, vc.visited[target.StringCanonical()]...)
-		if prevLenList == len(list) {
-			// The list we have now is the same we just checked if it's done or waiting on us.
-			// We are finished.
-			return dependents, nil // no unlocking on purpose
-		}
-		prevLenList = len(list)
-		vc.mu.Unlock()
-		// Wait for sts's to be done outside of the mu lock.
-	stsLoop:
-		for _, sts := range list {
-			if dependents[sts.ID] {
-				// No need to wait if it's a dependent, because the sts is waiting on us.
-				// It's safe to perform comparison if they are waiting on us.
-				continue
-			}
-			for {
-				select {
-				case <-ctx.Done():
-					return nil, ctx.Err()
-				case newID := <-parentDepSub:
-					dependents[newID] = true
-					if newID == sts.ID {
-						// Just the one we were waiting for. It seems that it is
-						// now waiting on us.
-						continue stsLoop
-					}
-				case <-sts.Done():
-					continue stsLoop
-				}
-			}
-		}
-	}
-}
-
-// CompareTargetInputs compares two targets and their inputs to check if they are the same.
-func CompareTargetInputs(target domain.Target, platr *platutil.Resolver, allowPrivileged bool, overridingVars *variables.Scope, other dedup.TargetInput) (bool, error) {
-	if target.StringCanonical() != other.TargetCanonical {
-		return false, nil
-	}
-	if allowPrivileged != other.AllowPrivileged {
-		return false, nil
-	}
-	stsPlat, err := platr.ParseAllowNativeAndUser(other.Platform)
-	if err != nil {
-		return false, err
-	}
-	if !platr.PlatformEquals(platr.Current(), stsPlat) {
-		return false, nil
-	}
-	for _, bai := range other.BuildArgs {
-		variable, found := overridingVars.Get(bai.Name)
-		if found {
-			baiVariable := dedup.BuildArgInput{
-				Name:          bai.Name,
-				DefaultValue:  bai.DefaultValue,
-				ConstantValue: variable,
-			}
-			if !baiVariable.Equals(bai) {
-				return false, nil
-			}
-		} else {
-			if !bai.IsDefaultValue() {
-				return false, nil
-			}
-		}
-	}
-	return true, nil
+type VisitedCollection interface {
+	All() []*SingleTarget
+	Add(ctx context.Context, target domain.Target, platr *platutil.Resolver, allowPrivileged bool, overridingVars *variables.Scope, parentDepSub chan string) (*SingleTarget, bool, error)
```

Comment:
Mind copying the documentation of these functions + add documentation for the interface itself?